### PR TITLE
feat: Allow ad hoc sharing of existing case (FPLA-2138)

### PR DESCRIPTION
**Pull request by tomaszpowroznik**
https://tools.hmcts.net/jira/browse/FPLA-2138

https://tools.hmcts.net/confluence/display/RPL/Share+existing+case

### CODE REVIEW

File: scripts/shareCase.sh [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+
+buildPayload() {
+    usersIdsArray=(${1//,/ })
+    usersIds=$(printf ",\"%s\"" "${usersIdsArray[@]}")
+    usersIds=${usersIds:1}
+    requestBody="{\"ids\":[${usersIds}]}"
+    echo ${requestBody}
+}
+
+getAccessToken() {
```
Comment (lewisbirks) [inline scripts/shareCase.sh:39]: Can we not use  `fpla-docker/bin/utils/idam-authenticate.sh` folder to get the auth token?

Comment (tomaszpowroznik) [reply scripts/shareCase.sh:39]: this is supposed to be standalone script, dont want devops to checkout any repo which contains dependencies

Comment (lewisbirks) [reply scripts/shareCase.sh:39]: Fair enough, thought that might be the case but wanted to raise it anyway

**Review by lewisbirks — COMMENTED** [review]
@akomisarek or @nkoranne do you mind looking over the script, I am happy with the rest of it just want someone else to look over the script as whilst I think it is fine I am not as confident in my scripting as I am java

File: service/src/main/java/uk/gov/hmcts/reform/fpl/controllers/support/ShareCaseController.java [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+    private final CaseUserApi caseUser;
+    private final AuthTokenGenerator authTokenGenerator;
+
+    @PostMapping(value = "/support/case/{caseId}/share", consumes = APPLICATION_JSON_VALUE)
+    @Secured("caseworker-publiclaw-systemupdate")
+    public void shareCase(@PathVariable("caseId") String caseId, @RequestBody Users users) {
+
+        final String serviceToken = authTokenGenerator.generate();
+
+        users.ids.forEach(
```
[the remark this answers is not in the document]
Comment (nkoranne) [reply service/src/main/java/uk/gov/hmcts/reform/fpl/controllers/support/ShareCaseController.java:39]: checked the script and it looks fine to me.

[the remark this answers is not in the document]
Comment (tomaszpowroznik) [reply service/src/main/java/uk/gov/hmcts/reform/fpl/controllers/support/ShareCaseController.java:39]: It is adhoc tool, performance should not be key consideration here. With concurrent execution (dozens of concurrent calls can choke ccd up). Ideally should use new bulk ccd endpoint

[the remark this answers is not in the document]
Comment (akomisarek) [reply service/src/main/java/uk/gov/hmcts/reform/fpl/controllers/support/ShareCaseController.java:39]: Tomek your thoughts?

File: service/src/integrationTest/java/uk/gov/hmcts/reform/fpl/controllers/support/ShareCaseControllerTest.java [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+            new CaseUser(userId, Set.of("[CREATOR]", "[LASOLICITOR]")));
+    }
+
+    private MvcResult shareCase(final String caseId, List<String> userIds) throws Exception {
+        final String body = format("{%s:[%s]}",
+            quote("ids"),
+            userIds.stream().map(this::quote).collect(joining(",")));
+
+        return shareCase(caseId, body);
+    }
```
Comment (akomisarek) [inline service/src/integrationTest/java/uk/gov/hmcts/reform/fpl/controllers/support/ShareCaseControllerTest.java:106]: I believe using ObjectMapper would be nicer.

**Review by akomisarek — DISMISSED** [review]
@lewisbirks I think all of us are googlebash level :D

**Review by akomisarek — APPROVED** [review]

**Pull request merged**